crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -107,8 +107,53 @@ def cancel
         }
       end
 
+      # Orchestrate the upgrade of the nodes
+      def nodes
+        # check for current global status
+        # 1. TODO: return if upgrade has finished
+        # 2. TODO: find the next big step
+        next_step = "controllers"
+
+        if next_step == "controllers"
+
+          # TODO: Save the "current_step" to global status
+          if upgrade_controller_nodes
+            # upgrading controller nodes succeeded, we can continue with computes
+            next_step = "computes"
+          else
+            # upgrading controller nodes has failed, exiting
+            # leaving next_step as "controllers", so we continue from correct point on retry
+            return false
+          end
+        end
+
+        if next_step == "computes"
+          # TODO: Save the "current_step" to global status
+          upgrade_compute_nodes
+        end
+        true
+      end
+
       protected
 
+      def upgrade_controller_nodes
+        # TODO: find the controller node that needs to be upgraded now
+        # First node to upgrade is DRBD slave
+        node = NodeObject.find("drbd_rsc_postgresql_master:false").first
```

Comment:
No, [as I mentioned previously](https://github.com/crowbar/crowbar-core/pull/655#discussion_r79941960), there is no guarantee that this Chef data really reflects the master/slave status.  [Chef actually tracks whether the node is the founder](https://github.com/crowbar/crowbar-ha/blob/master/chef/cookbooks/crowbar-pacemaker/providers/drbd.rb#L36), not whether it's the master.  You should query Pacemaker to find which is the master.  An XPath query on `cibadmin -Ql` is probably the most correct, although parsing `crm resource status ms-drbd-{postgresql,rabbitmq}` is easier and probably good enough.